Rewrite Redis manager

kminehart opened the pull request:
1. I really did not like the Redis implementation that existed; not only because it was half-finished, but because it dropped every policy into a single hashmap key rather than taking advantage of the key/value storage aspect of Redis.

2. I finished the unimplemented functions

3. There was a really bad memory leak in `redisUnmarshalPolicy`. I'm still struggling to find out why exactly but my implementation seemed to fix this.

4. I added some basic tests just for sanity.

Lastly, if `ladon.DefaultPolicy` implemented [`BinaryMarshaler`](https://golang.org/pkg/encoding/#BinaryMarshaler) and [`BinaryUnmarshaler`](https://golang.org/pkg/encoding/#BinaryUnmarshaler) then we could avoid a lot of to-and-from JSON overhead.

I think Redis is honestly the best candidate for a storage backend for Ladon. We ran into tons of performance issues with multiple unique permission checks when using Postgres. In my opinion, Redis should have more official support.

kminehart committed:
ory-am -> ory

kminehart committed:
update function signatures

kminehart committed:
rebuild the redis manager

kminehart committed:
Add extra tests, fix update and delete

kminehart committed:
add extra test

aeneasr commented:
Awesome! I don't really have any redis knowledge so it will be hard for me to review. One question is if this breaks backwards compatibility? I'll also try to activate tests so we can see if the test suite works!

kminehart committed:
Huge improvements to FindRequestCandidates performance

kminehart commented:
Correct, this is not backwards compatible at all.

I figured that wouldn't be an issue considering only half of the functions were implemented, and with the memory leak I found, I can't imagine this was actually being used in a production setting. Also, it wouldn't even compile because of the references to `ory-am` instead of `ory`

Also: I just fixed an issue where, like the memory manager (https://github.com/ory/ladon/tree/master/manager/memory) `FindRequestCandidates` just returns every policy. This can start to get a little ridiculous when you have thousands of policies.

Now, policies are stored both individually in keys with `SET`, and in hashmaps. Each Subject and Resource is a hashmap, which contain policies that apply to them.

So now, `FindRequestCandidates` should just give you a list of policies that apply to the requested resource and subject. This increases the complexity of creating a deleting policies, but reduces the complexity of `CanAccess`.

---

Edit: One issue I see with this now is subjects with regexes.

A policy like this, to allow all users to read all profiles, would fail.

```
Subject: profiles:<.*>
Resources: user:<.*>
Action: read
Effect: allow
```

 would not be found. :thinking: 

---

I added a test case for this specific problem. How do you solve this in the postgres manager, @arekkas ? I imagine the solution would be similar.

Looking at your implementation of the database manager, it looks like you're letting the database evaluate the regex. Unfortunately, a full regex implementation doesn't exist in Redis, so I'm not sure there's an efficient way to actually fix this. It only really supports globbing (https://redis.io/commands/scan)

I could read all of the keys and evaluate the regular expressions in the library, but that would take an enormous amount of time. If the complexity of `FindRequestCandidates` is too high, it could take too much time to return a response from `CanAccess`, which was why we switched to Redis in the first place.

I'd honestly be happy saying that 1 disadvantage / requirement of using this Redis manager is that in each policy, either the subject(s) or resource(s) must be without a regular expression.

So for that policy above:

```
Subjects: profiles:<.*>
Resources: group:profile-reader
Actions: read
Effect: allow
```

and the request would be:

```
Subject: profile:xyz
Resource: group:proflie-reader
Action: read
```

would be fine (and honestly preferred).

Also, this repository needs a license.

aeneasr commented:
> I figured that wouldn't be an issue considering only half of the functions were implemented, and with the memory leak I found, I can't imagine this was actually being used in a production setting. Also, it wouldn't even compile because of the references to ory-am instead of ory 😂

That makes sense :)

> Looking at your implementation of the database manager, it looks like you're letting the database evaluate the regex. Unfortunately, a full regex implementation doesn't exist in Redis, so I'm not sure there's an efficient way to actually fix this. It only really supports globbing (https://redis.io/commands/scan)

That's not a problem, ladon re-evaluates all policies anyways so the regexp matcher is independent of the DBAL. In the case of postgres however, it did make sense to have the regexp there too!

> Also, this repository needs a license. 👀

Oh yeah, absolutely - I'll add it right now.

kminehart commented:
Gotya. Let me know if you need me to change anything.

aeneasr commented:
Thank you for your contribution!
